### ISSUE TITLE
fetchOrder signature

### ISSUE THREAD

**Issue by wannesdemaeght** [ccxt/ccxt#1484]
> The signature of the fetchOrder/fetch_order method is as follows:
> 
> //  you can use the params argument for custom overrides
> exchange.fetchOrder (id, symbol = undefined, params = {})

Since I've implemented fetchOrder for Kucoin, this is no longer true.
(id, symbol = undefined, type = undefined, params = {})

Passing this signature to other exchanges' fetchOrder gives mixed results.
Gdax ignores the unneeded statements and works ok.
Binance did not ignore the extra statements, and had to adapt the code.

**Comment by wannesdemaeght** [comment]
I don't quite follow.
Unfortunately, all 3 parameters are mandatory for kucoin.
Passing the third one via params{} would also not be ideal (because that would also violate the signature for other exchanges)

How are you going to pass the type to kucoin,  if there isn't a parameter for it?

**Comment by kroitor** [comment]
@wannesdemaeght according to their docs, type isn't required. So, this is what we do everywhere by default – omit it. And if the user needs a specific side, he will use the extra `params`, but in most cases the user wants all of them (both `BUY` and `SELL` sides). Does this make sense?

**Comment by wannesdemaeght** [comment]
No, type IS required unfortunately (try without, exchange returns an error that it is required)

https://kucoinapidocs.docs.apiary.io/#reference/0/trading/order-details
Required are symbol, type, and orderOid.

**Comment by kroitor** [comment]
Oh, if you mean fetchOrder only, yes, for it the type is required, and in that case we have two options:

1. either make a non-standard kucoin-specific signature
2. use the extra params

We prefer the second way. Otherwise it would require hardcoding specific codepaths for kucoin.

**Comment by wannesdemaeght** [comment]
would it not be easier to just change the default signature for fetchOrder to 
(id, symbol = undefined, type = undefined, params = {}) ?

Here's why:
because of unified response and unified call methods, I can do stuff like this:
```
const responses = await Promise.all([
	exch[open[i].sell].fetchOrder(open[i].ID.sell, open[i].market, "SELL"),
	exch[open[i].buy].fetchOrder(open[i].ID.buy, open[i].market, "BUY"),
]);
const resultSell	= responses[0];
const resultBuy 	= responses[1];
```
Where I just have a list of Buy-orderOids, and accompanying Sell-orderOids.

**Comment by kroitor** [comment]
> would it not be easier to just change the default signature for fetchOrder to
(id, symbol = undefined, type = undefined, params = {}) ?

For one exchange of 90? Don't know yet )

**Comment by kroitor** [comment]
Wouldn't it be easier to write:
```JavaScript
const responses = await Promise.all([
	exch[open[i].sell].fetchOrder(open[i].ID.sell, open[i].market, sellParams),
	exch[open[i].buy].fetchOrder(open[i].ID.buy, open[i].market, buyParams),
]);
const resultSell	= responses[0];
const resultBuy 	= responses[1];
```

?

**Comment by wannesdemaeght** [comment]
no, because buyParams and sellParams would get passed regardless of the exchange.
Binance for instance would not like the param type=Buy

**Comment by kroitor** [comment]
```JavaScript

buyParamsPerExchange = {
    'binance': { }, // look ma, no side here )
    'kucoin': { 'type': 'BUY' },
}
sellParamsPerExchange = {
    'binance': { }, // look ma, no side here )
    'kucoin': { 'type': 'SELL' },
}

```

No? )

**Comment by wannesdemaeght** [comment]
ok, that would work, but right now, each exchange is set to pass ALL params regardless.

**Comment by kroitor** [comment]
@wannesdemaeght I mean, it really depends on the distinction between a library-scope task and a userland-scope task, and we believe differentiating between types of orders and organizing the params accordingly is a userland task. Otherwise we would be hunting every possible use-case, and that would quickly render the entire library unusable. It's the "design guard", not the convenience issue. The ultimate goal of the library is to be as thin wrapper as possible, and if we account for all their quirks and minor differences, this would hardly be possible. We would add `type` if more than 50% of major exchanges required it, but they don't. Does this make sense now? )

**Comment by wannesdemaeght** [comment]
I see your point, but would like to point out that adding "type"to the signature is just an empty placeholder.
No harm would come of it, and it would not bloat anything.
It would also be a quick fix, which I would be happy to do for you.

But, your proposed method would also work on my end, so I will go that way if that is more viable.
Thank you for taking the time for answering this thoroughly!

(probably will need to document the need for adding "type=BUY" in the parameters for kucoin for other users)
